tuxedo-core/src/verifier/htlc.rs

```diff
@@ -0,0 +1,380 @@
+/// namely the current block number. So there is a decision to be made:
+/// * Allow the verifier to take come config and grab that info by calling a function given in the config.
+///   This is what we do with constraint checker.
+/// * Modify the `Verifier` trait to pass along the block number.
+///
+/// On the one hand the block number seems like a pretty fundamental and basic thing to add. On the other
+/// hand there could be many more things to pass. For example, the timestamp.
+/// However any more complex information would require coupling with Constraint Checkers and it is not
+/// easy to red state like in accounts.
+///
+/// Decision: I will add block number to the signature. And remain open to adding more blockchain-level
+/// fundamental things. Especially if they are available in bitcoin script.
+///
+/// Regarding the verifier constraint checker separation, perhaps the right line to be drawn is
+/// that verifiers are useful in a lot of places, but perhaps not expressive enough in others.
+/// When they are not expressive enough, just use `UpForGrabs` and rely on the constraint checker,
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct TimeLock {
+    pub unlock_block_height: u32,
+}
+
+impl Verifier for TimeLock {
+    type Redeemer = ();
+    fn verify(&self, _: &[u8], block_height: u32, _: &()) -> bool {
+        block_height >= self.unlock_block_height
+    }
+}
+
+/// Allows UTXOs to be spent when a preimage to a recorded hash is provided.
+/// This could be used as a puzzle (although a partial preimage search would be better)
+/// or a means of sharing a password, or as part of a simple atomic swapping protocol.
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct BlakeTwoHashLock {
+    pub hash_lock: H256,
+}
+
+impl BlakeTwoHashLock {
+    pub fn new_from_secret<T: Encode>(secret: T) -> Self {
+        Self {
+            hash_lock: BlakeTwo256::hash(&secret.encode()),
+        }
+    }
+}
+
+impl Verifier for BlakeTwoHashLock {
+    type Redeemer = Vec<u8>;
+    fn verify(&self, _: &[u8], _: u32, secret: &Self::Redeemer) -> bool {
+        BlakeTwo256::hash(secret) == self.hash_lock
+    }
+}
+
+/// Allows a UTXO to be spent, and therefore acknowledged by an intended recipient by revealing
+/// a hash preimage. After an initial claim period elapses on chain, the UTXO can also be spent
+/// by the refunder. In practice, the refunder is often the same address initially funded the HTLC.
+///
+/// The receiver and refunder are specified as a simple public keys for simplicity. It would be
+/// interesting to use public key hash, or better yet, simply abstract this over some opaque
+/// inner verifier for maximum composability.
+///
+/// @lederstrumpf when the time expires, is the primary "happy" path supposed to remain open?
+/// Or is it that if the swap hasn't started on time, the refund is the only option.
```

Comment:
I think these kinds of comments should be kept out of the docs

Reply:
Agreed. This one needs to be resolved before this is merged. Based Andreas's point in the video ([link to exact moment](https://youtu.be/8FeAXjkmDcQ?si=RkVmBBbaXo84WwAN&t=1492)) I believe that that the happy path should remain enabled which is how I have it now.

Reply:
> /// @lederstrumpf when the time expires, is the primary "happy" path supposed to remain open?
> /// Or is it that if the swap hasn't started on time, the refund is the only option.

Strong question you raise.
In the Farcaster protocol, the happy path closes once one of the parties initiates the cancel path.

HTLC atomic swap protocols on Bitcoin keep the happy path open even after timeout expires - see e.g. the COMIT spec: https://comit.network/docs/0.13.0/core-concepts/atomic-swap-htlc/#htlcs-on-different-ledgers (interestingly, they've done the same for the ETH protocol: https://github.com/comit-network/RFCs/blob/master/RFC-007-SWAP-Basic-Ether.adoc)
or the original proposal:
https://bitcointalk.org/index.php?topic=193281.msg2224949#msg2224949
or BIP 119:
https://github.com/bitcoin/bips/blob/master/bip-0199.mediawiki#summary

The reason for this is indeed that there's deliberately no `OP_CHECKSEQUENCE` nor `OP_CHECKLOCKTIME` - only their guardian variants `OP_CHECKSEQUENCEVERIFY` & `OP_CHECKLOCKTIMEVERIFY`.
Most discussion around this is probably in the annals of the bitcoin wizards IRC, if you wanna investigate deeper.
The arguments that have bubbled to my surface:
Satoshi's stance in 2010 (i.e. The Prophet's unassailable proclamation in Bitcoin Land, for better or worse): https://bitcointalk.org/index.php?topic=1786.msg22119#msg22119
Jeremy Rubin's AMA: https://www.reddit.com/r/Bitcoin/comments/gwjr8p/comment/ft80amd/?utm_source=share&utm_medium=web2x&context=3
Another disgruntled (& highly ignored) seeker of The Truth's summary: https://bitcoin.stackexchange.com/questions/96366/what-are-the-reasons-to-avoid-spend-paths-that-become-invalid-over-time-without

Now, while I have reservations about the reorg arguments even in Bitcoin's context, they are not applicable to chains with deterministic finality, i.e. all substrate based chains (aside from Kulupu).
i.e. with deterministic finality, you could have the equivalent of `OP_CHECK{SEQUENCE,LOCKTIME}`, and it would be safe as long as the output being spent has been finalized.

I reckon it might be preferable to have mutually exclusive success & refund paths for HTLC based swaps too - with a more general definition of HTLC, i.e. going beyond what Bitcoin Script allows.
It would still not ensure atomicity (since the chains run on different clocks), but improves safety of the swap for Bob assuming chain A's clock doesn't outrace chain B's clock.
Breakable assumption if crosschain: `t1 < t2`:
Let's take chain A to be the chain where Alice funds TX1 (with a timelock `t2` for her refund `TX2`) and chain B as the chain where Bob funds `TX3` (with a timelock `t1` for his refund `TX4`)

Assuming both funding txs have been published, but Alice has not claimed yet, then the possible spending conditions are:

`TX1` (Chain A): `(x ∧ Bob) ∨ (t2 ∧ Alice)`
`TX3` (Chain B): `(x ∧ Alice) ∨ (t1 ∧ Bob)`

In detail:
- `!t1 ∧ !t2`: Alice can claim. Bob can't do jack and has to wait for `t1` or Alice to claim.
- `t1 ∧ !t2` (`t1` has passed): Alice can claim (unsafe since Bob would learn `x` and his refund might frontrun her claim tx) or wait for `t2` to refund. Bob can (and should) refund or (unsafe!) wait for Alice to claim.
- `t1 ∧ t2` (both `t1` and `t2` have passed): Alice can claim or refund. Bob can refund and should already have done so before `t2` passed, else non-atomic.
- `!t1 ∧ t2` (`t2` has passed): Alice can claim or refund. Bob has to wait for `t1` or for Alice to claim: non-atomic scenario.

If you now impose that the happy path is unavailable after `t1` or `t2` respectively, then this changes to:
Chain A: `(!t2 ∧ x ∧ Bob) ∨ (t2 ∧ Alice)`
Chain B: `(!t1 ∧ x ∧ Alice) ∨ (t1 ∧ Bob)`

- `!t1 ∧ !t2`: Alice can claim. Bob can't do jack and has to wait for `t1` or for Alice to claim.
- `t1 ∧ !t2` (`t1` has passed): Alice can't claim (which would anyway be unsafe), so must wait for `t2` to refund. Bob can refund.
- `t1 ∧ t2` (both `t1` and `t2` have passed): Alice can only refund. Bob can only refund.
- `!t1 ∧ t2` (`t2` has passed): Alice can claim or refund. Bob has to wait for `t1` or for Alice to claim: non-atomic scenario.

The primary advantage is that Alice can't steal from Bob unless she publishes `TX2` prior to `t1`.
Note you'd get the same if you'd just make the paths on `TX1` temporally mutually exclusive, assuming Alice never leaks `x` to Bob (e.g. by hopelessly trying to spend `TX1`, or by getting hacked).
So if this was a crosschain swap of Tuxedo with e.g. Bitcoin, you could retain this advantage as long as `Tuxedo = Chain A`.

Reply:
TLDR: to keep compatibility with Bitcoin Script (not in terms of only supporting its features, but also the *assumptions* about what can happen), I'd keep the happy path open.

But keeping compatibility with BS's assumptions means restricting `Verifier` rules to what BS would permit, so this is a pretty fundamental decision. The other alternative is going with whatever makes most sense for a particular protocol and breaking compatibility with BS's assumptions. And that's also a pretty fundamental decision ;)

Reply:
Because Tuxedo is a framework, we don't have to make any decision once and for all here. The framework can support any old verifier whether it is compatible with bitcoin or not.

When it comes time to launch a Tuxedo-based chain then we have to decide what verifiers are concretely allowed. I don't think we should enforce bitcoin compatibility to the entire Tuxedo ecosystem.